transport.go

```diff
@@ -202,6 +219,37 @@ func (t *transport) Dial(ctx context.Context, raddr ma.Multiaddr, p peer.ID) (tp
 	return conn, nil
 }
 
+func (t *transport) holePunch(network string, addr *net.UDPAddr) error {
+	pconn, err := t.connManager.Dial(network, addr)
+	if err != nil {
+		return err
+	}
+	defer pconn.DecreaseCount()
+
+	conn := pconn.UDPConn
+
+	payload := make([]byte, 64)
+	_, err = rand.Read(payload)
+	if err != nil {
+		return err
+	}
+
+	_, err = conn.WriteToUDP(payload, addr)
+	if err != nil {
+		return err
+	}
+
+	for i := 0; i < 10; i++ {
```

Comment:
10 seems excessive. Maybe reduce it to 3?

Reply:
Or, alternatively, should we build some cancellation logic? Something like a `map[*net.UDPAddr]context.CancelFunc activeHolePunches`, and every time `Listen` accepts a new connection it would call the cancel function if we're currently punching holes for this remote address.

Reply:
Yeah, agreed, 10 might be a bit excessive, but I want to keep punching for a reasonable amount of time (say 100ms) until the conn is there or there is some form of timeout.

And yes indeed, having listen interrupt the hole punch if it accepts a connection from the target address is ideal.
It might be a little complex to do so, but it's the _right_ thing to do.

Reply:
Ok, let's do the easy thing in this PR and keep the optimization for a later time. I'd be happy to help with it btw.

Reply:
thanks -- I'll check if it is reasonably easy to do right now, otherwise I will add a TODO.

Reply:
It seems it's pretty easy to do, if we track by peer ID -- we just hook into the listener accept and cancel all the hole punching attempts for that peer.

So let's implement it, on it!

Reply:
ok, I've implemented tracking with a `map[peer.ID][*net.UDPAddr]` and the listener cancels all pending hole punches when it gets a connection from the peer.

Only thing that remains is the rate (and amount of packets) that we send.
I have set a timeout of 1s and do `10ms + rand((i+1)*10*ms` for the wait.
We might want to tune this if it's too many packets.